tools/targets/__init__.py

```diff
@@ -125,18 +126,23 @@ def get_resolution_order(json_data, target_name, order, level=0):
 
 def target(name, json_data):
     """Construct a target object"""
+    if name.startswith("_"):
+        raise Exception("Invalid target name '%s' specified, target name should not start with '_'" % name)
+    
     try:
         resolution_order = get_resolution_order(json_data, name, [])
     except KeyError as exc:
         raise_from(NotSupportedException(
             "target {} has an incomplete target definition".format(name)
         ), exc)
     resolution_order_names = [tgt for tgt, _ in resolution_order]
+    
     return Target(name=name,
                   json_data={key: value for key, value in json_data.items()
                              if key in resolution_order_names},
                   resolution_order=resolution_order,
-                  resolution_order_names=resolution_order_names)
+                  resolution_order_names=resolution_order_names,
+                  build_tools_metadata=json_data.get("__build_tools_metadata__", default_build_tools_metadata))
```

Comment:
It seems odd to have to add the metadata to each target, vs having a singleton of the data somewhere. I don't really have a better suggestion at the moment (besides maybe a dedicated function that fetches it?). Maybe @theotherjimmy has a better idea.

Reply:
I disagree, as the online compiler will use the same process to build multiple versions of mbed-os. That would require the singleton to have different values, and break your assumption that it can be a singleton. This is not "target data" but it is data we need for every target we compile.

Reply:
@bridadan In general, we can't have singletons that could take more than one value, as the online compiler will break the assumption that it can be a singleton in those cases. Things like "what version of ARM architecture does this use" are not going to change, and therefore can be "singletons". Most people call them constants though.